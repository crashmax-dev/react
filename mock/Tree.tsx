import React from 'react'

export function App(props: React.PropsWithChildren) {
  return <div aria-label="app">{props.children}</div>
}

export function Layout(props: React.PropsWithChildren) {
  return <div aria-label="layout">{props.children}</div>
}

type StateProps = React.PropsWithChildren<{
  initialValue: string
}>

export function State(props: StateProps) {
  return (
    <div aria-label="state">
      <h1 aria-label="state-initial-value">{props.initialValue}</h1>
      {props.children}
    </div>
  )
}
