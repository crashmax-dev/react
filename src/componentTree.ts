import { createElement } from 'react'

type Provider<Props = {}> = [React.ElementType, Props?]

export function componentTree(providers: Provider[]) {
  return ({ children }: { children: JSX.Element }) => {
    const lastIndex = providers.length - 1
    let currentChildren = children

    for (let i = lastIndex; i >= 0; i--) {
      const [component, props = {}] = providers[i]!
      currentChildren = createElement(component, props, currentChildren)
    }

    return currentChildren
  }
}
