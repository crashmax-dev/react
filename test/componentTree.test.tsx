import { App, Layout, State } from '../mock/Tree.js'
import { componentTree } from '../src/componentTree.js'
import { render, screen } from './utils.js'

describe('componentTree', () => {
  it('should be defined', () => {
    expect(componentTree).toBeDefined()
  })

  it('should be render component tree', () => {
    const Tree = componentTree([[State, { initialValue: 'hello' }], [Layout]])

    render(
      <Tree>
        <App />
      </Tree>
    )
    expect(screen.getByLabelText('state')).toMatchSnapshot()
  })
})
