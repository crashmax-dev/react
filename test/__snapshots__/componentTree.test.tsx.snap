// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`componentTree > should be render component tree 1`] = `
<div
  aria-label="state"
>
  <h1
    aria-label="state-initial-value"
  >
    hello
  </h1>
  <div
    aria-label="layout"
  >
    <div
      aria-label="app"
    />
  </div>
</div>
`;
